File: bench-the-rebuild/ink-and-tsv/go/k5/main.go
// K5 — TSV parsing, Go port. Same three kernels and the same canonical round-trip checksums as the
// Rust port and the Python original.
package main

import (
	"crypto/sha256"
	"fmt"
	"os"
	"strconv"
	"sync"
	"time"
)

type Row struct {
	Codepoints []uint32
	Glyphs     []string
	Clusters   []int64
	Seams      []string
	Positions  [][3]int64
}

type AuditRow struct {
	Config       string
	Codepoints   string
	Kinds        []string
	MatchedEntry string
	Baseline     []string
	New          []string
}

func parseHex(s string) uint32 {
	var v uint32
	for i := 0; i < len(s); i++ {
		c := s[i]
		switch {
		case c >= '0' && c <= '9':
			v = v*16 + uint32(c-'0')
		case c >= 'a' && c <= 'f':
			v = v*16 + uint32(c-'a'+10)
		default:
			v = v*16 + uint32(c-'A'+10)
		}
	}
	return v
}

func parseInt(s string) int64 {
	neg := false
	if len(s) > 0 && s[0] == '-' {
		neg = true
		s = s[1:]
	}
	var v int64
	for i := 0; i < len(s); i++ {
		v = v*10 + int64(s[i]-'0')
	}
	if neg {
		return -v
	}
	return v
}

func splitN(s string, sep byte, n int) []string {
	out := make([]string, 0, n)
	for {
		if len(out) == n-1 {
			out = append(out, s)
			return out
		}
		i := indexByte(s, sep)
		if i < 0 {
			out = append(out, s)
			return out
		}
		out = append(out, s[:i])
		s = s[i+1:]
	}
}

func indexByte(s string, c byte) int {
	for i := 0; i < len(s); i++ {
		if s[i] == c {
			return i
		}
	}
	return -1
}

func splitAll(s string, sep byte) []string {
	n := 1
	for i := 0; i < len(s); i++ {
		if s[i] == sep {
			n++
		}
	}
	out := make([]string, 0, n)
	start := 0
	for i := 0; i < len(s); i++ {
		if s[i] == sep {
			out = append(out, s[start:i])
			start = i + 1
		}
	}
	return append(out, s[start:])
}

func parseRows(text string) []Row {
	out := make([]Row, 0, 65536)
	for _, line := range splitAll(text, '\n') {
		if len(line) == 0 || line[0] == '#' {
			continue
		}
		f := splitN(line, '\t', 5)
		cps := splitAll(f[0], ':')
		codepoints := make([]uint32, len(cps))
		for i, c := range cps {
			codepoints[i] = parseHex(c)
		}
		cl := splitAll(f[2], ',')
		clusters := make([]int64, len(cl))
		for i, c := range cl {
			clusters[i] = parseInt(c)
		}
		var seams []string
		if f[3] != "" {
			seams = splitAll(f[3], ',')
		}
		ps := splitAll(f[4], '|')
		positions := make([][3]int64, len(ps))
		for i, p := range ps {
			t := splitN(p, ',', 3)
			positions[i] = [3]int64{parseInt(t[0]), parseInt(t[1]), parseInt(t[2])}
		}
		out = append(out, Row{codepoints, splitAll(f[1], '|'), clusters, seams, positions})
	}
	return out
}

func rowsChecksum(rows []Row) string {
	h := sha256.New()
	buf := make([]byte, 0, 4096)
	for _, r := range rows {
		buf = buf[:0]
		for i, cp := range r.Codepoints {
			if i > 0 {
				buf = append(buf, ':')
			}
			buf = append(buf, fmt.Sprintf("%04X", cp)...)
		}
		buf = append(buf, '\t')
		for i, g := range r.Glyphs {
			if i > 0 {
				buf = append(buf, '|')
			}
			buf = append(buf, g...)
		}
		buf = append(buf, '\t')
		for i, c := range r.Clusters {
			if i > 0 {
				buf = append(buf, ',')
			}
			buf = strconv.AppendInt(buf, c, 10)
		}
		buf = append(buf, '\t')
		for i, s := range r.Seams {
			if i > 0 {
				buf = append(buf, ',')
			}
			buf = append(buf, s...)
		}
		buf = append(buf, '\t')
		for i, p := range r.Positions {
			if i > 0 {
				buf = append(buf, '|')
			}
			buf = strconv.AppendInt(buf, p[0], 10)
			buf = append(buf, ',')
			buf = strconv.AppendInt(buf, p[1], 10)
			buf = append(buf, ',')
			buf = strconv.AppendInt(buf, p[2], 10)
		}
		buf = append(buf, '\n')
		h.Write(buf)
	}
	return fmt.Sprintf("%x", h.Sum(nil))
}

func parseAudit(text string) []AuditRow {
	lines := splitAll(text, '\n')
	out := make([]AuditRow, 0, 300000)
	for _, line := range lines[1:] {
		if len(line) == 0 {
			continue
		}
		f := splitN(line, '\t', 6)
		out = append(out, AuditRow{f[0], f[1], splitAll(f[2], ','), f[3], splitAll(f[4], '|'), splitAll(f[5], '|')})
	}
	return out
}

func auditChecksum(rows []AuditRow) string {
	h := sha256.New()
	buf := make([]byte, 0, 1024)
	for _, r := range rows {
		buf = buf[:0]
		buf = append(buf, r.Config...)
		buf = append(buf, '\t')
		buf = append(buf, r.Codepoints...)
		buf = append(buf, '\t')
		for i, k := range r.Kinds {
			if i > 0 {
				buf = append(buf, ',')
			}
			buf = append(buf, k...)
		}
		buf = append(buf, '\t')
		buf = append(buf, r.MatchedEntry...)
		buf = append(buf, '\t')
		for i, b := range r.Baseline {
			if i > 0 {
				buf = append(buf, '|')
			}
			buf = append(buf, b...)
		}
		buf = append(buf, '\t')
		for i, n := range r.New {
			if i > 0 {
				buf = append(buf, '|')
			}
			buf = append(buf, n...)
		}
		buf = append(buf, '\n')
		h.Write(buf)
	}
	return fmt.Sprintf("%x", h.Sum(nil))
}

var alphabet = map[uint32]bool{
	0x0020: true, 0x00B7: true, 0x200C: true, 0xE650: true, 0xE652: true, 0xE653: true,
	0xE658: true, 0xE65A: true, 0xE665: true, 0xE666: true, 0xE667: true, 0xE668: true,
	0xE670: true, 0xE672: true, 0xE675: true, 0xE676: true, 0xE679: true, 0xE67A: true,
}

func inAlphabet(tok []byte) bool {
	var v uint32
	for _, c := range tok {
		switch {
		case c >= '0' && c <= '9':
			v = v*16 + uint32(c-'0')
		case c >= 'a' && c <= 'f':
			v = v*16 + uint32(c-'a'+10)
		case c >= 'A' && c <= 'F':
			v = v*16 + uint32(c-'A'+10)
		default:
			return false
		}
	}
	return alphabet[v]
}

func keepLine(line []byte) bool {
	end := len(line)
	for i := 0; i < len(line); i++ {
		if line[i] == '\t' {
			end = i
			break
		}
	}
	field := line[:end]
	start := 0
	for i := 0; i <= len(field); i++ {
		if i == len(field) || field[i] == ':' {
			if !inAlphabet(field[start:i]) {
				return false
			}
			start = i + 1
		}
	}
	return true
}

func filterChunk(data []byte, out []byte) ([]byte, int) {
	kept := 0
	start := 0
	for i := 0; i <= len(data); i++ {
		if i == len(data) || data[i] == '\n' {
			line := data[start:i]
			start = i + 1
			if len(line) == 0 {
				continue
			}
			if line[0] == '#' {
				out = append(out, line...)
				out = append(out, '\n')
				continue
			}
			if keepLine(line) {
				out = append(out, line...)
				out = append(out, '\n')
				kept++
			}
		}
	}
	return out, kept
}

func splitAtLines(data []byte, parts int) [][]byte {
	bounds := []int{0}
	for i := 1; i < parts; i++ {
		p := len(data) * i / parts
		for p < len(data) && data[p] != '\n' {
			p++
		}
		if p < len(data) {
			p++
		}
		if p > len(data) {
			p = len(data)
		}
		if p > bounds[len(bounds)-1] {
			bounds = append(bounds, p)
		}
	}
	bounds = append(bounds, len(data))
	out := make([][]byte, 0, len(bounds)-1)
	for i := 0; i < len(bounds)-1; i++ {
		out = append(out, data[bounds[i]:bounds[i+1]])
	}
	return out
}

func main() {
	rowsPath, auditPath, bigPath, outDir := os.Args[1], os.Args[2], os.Args[3], os.Args[4]
	threads := 8
	if len(os.Args) > 5 {
		threads, _ = strconv.Atoi(os.Args[5])
	}

	rowsRaw, _ := os.ReadFile(rowsPath)
	rowsText := string(rowsRaw)
	bestRows := 1e30
	var rows []Row
	for i := 0; i < 5; i++ {
		t := time.Now()
		rows = parseRows(rowsText)
		if e := time.Since(t).Seconds(); e < bestRows {
			bestRows = e
		}
	}
	ckRows := rowsChecksum(rows)
	nRows := len(rows)
	rows = nil

	auditRaw, _ := os.ReadFile(auditPath)
	auditText := string(auditRaw)
	bestAudit := 1e30
	var audit []AuditRow
	for i := 0; i < 3; i++ {
		t := time.Now()
		audit = parseAudit(auditText)
		if e := time.Since(t).Seconds(); e < bestAudit {
			bestAudit = e
		}
	}
	ckAudit := auditChecksum(audit)
	nAudit := len(audit)
	audit = nil

	big, _ := os.ReadFile(bigPath)
	bestFilter := 1e30
	var out []byte
	kept := 0
	for i := 0; i < 2; i++ {
		t := time.Now()
		buf := make([]byte, 0, 8<<20)
		buf, kept = filterChunk(big, buf)
		if e := time.Since(t).Seconds(); e < bestFilter {
			bestFilter = e
		}
		out = buf
	}
	ckFilter := fmt.Sprintf("%x", sha256.Sum256(out))
	os.WriteFile(outDir+"/go.subset.tsv", out, 0o644)
	out = nil

	bestPar := 1e30
	keptPar := 0
	var ckPar string
	for i := 0; i < 2; i++ {
		t := time.Now()
		chunks := splitAtLines(big, threads)
		parts := make([][]byte, len(chunks))
		counts := make([]int, len(chunks))
		var wg sync.WaitGroup
		for w := range chunks {
			wg.Add(1)
			go func(w int) {
				defer wg.Done()
				buf := make([]byte, 0, 2<<20)
				buf, counts[w] = filterChunk(chunks[w], buf)
				parts[w] = buf
			}(w)
		}
		wg.Wait()
		merged := make([]byte, 0, 8<<20)
		total := 0
		for w := range parts {
			merged = append(merged, parts[w]...)
			total += counts[w]
		}
		if e := time.Since(t).Seconds(); e < bestPar {
			bestPar = e
		}
		keptPar = total
		ckPar = fmt.Sprintf("%x", sha256.Sum256(merged))
	}

	fmt.Printf("{\"rows_from_tsv\":{\"rows\":%d,\"go_seconds\":%.6f,\"go_ns_per_row\":%.1f,\"go_checksum\":\"%s\"},"+
		"\"load_audit\":{\"rows\":%d,\"go_seconds\":%.6f,\"go_ns_per_row\":%.1f,\"go_checksum\":\"%s\"},"+
		"\"filter_table\":{\"source_bytes\":%d,\"go_single_seconds\":%.6f,\"go_single_kept\":%d,\"go_single_checksum\":\"%s\","+
		"\"go_parallel_threads\":%d,\"go_parallel_seconds\":%.6f,\"go_parallel_kept\":%d,\"go_parallel_checksum\":\"%s\"}}\n",
		nRows, bestRows, bestRows/float64(nRows)*1e9, ckRows,
		nAudit, bestAudit, bestAudit/float64(nAudit)*1e9, ckAudit,
		len(big), bestFilter, kept, ckFilter, threads, bestPar, keptPar, ckPar)
}

File: bench-the-rebuild/ink-and-tsv/rust/k5/src/main.rs
// K5 — TSV parsing, Rust port of rebuild/validation/rowmodel.py's Row.from_tsv,
// rebuild/review/audit.py's load_audit, and rebuild/pipeline/baseline_subset.py's filter_table.
//
// Equivalence: every variant is verified by re-serializing the parsed rows to the same canonical TSV
// the Python checksum uses and hashing that, so the printed checksum is a round trip through the
// parsed values and cannot be satisfied by echoing the input. The checksum pass runs outside the
// timer but consumes the whole parsed table, which is also what defeats dead-code elimination.

use sha2::{Digest, Sha256};
use std::time::Instant;

// ---- canonical re-serialization (must match k5_python.canon_row / canon_audit) -----------------

fn push_u32_hex4(out: &mut Vec<u8>, v: u32) {
    let s = format!("{:04X}", v);
    out.extend_from_slice(s.as_bytes());
}

fn push_i64(out: &mut Vec<u8>, v: i64) {
    let mut buf = itoa(v);
    out.append(&mut buf);
}

fn itoa(mut v: i64) -> Vec<u8> {
    if v == 0 {
        return vec![b'0'];
    }
    let neg = v < 0;
    let mut tmp = Vec::with_capacity(20);
    while v != 0 {
        tmp.push(b'0' + (v % 10).unsigned_abs() as u8);
        v /= 10;
    }
    if neg {
        tmp.push(b'-');
    }
    tmp.reverse();
    tmp
}

// ---- Row ---------------------------------------------------------------------------------------

struct RowOwned {
    codepoints: Vec<u32>,
    glyphs: Vec<String>,
    clusters: Vec<i64>,
    seams: Vec<String>,
    positions: Vec<(i64, i64, i64)>,
}

struct RowBorrowed<'a> {
    codepoints: Vec<u32>,
    glyphs: Vec<&'a str>,
    clusters: Vec<i64>,
    seams: Vec<&'a str>,
    positions: Vec<(i64, i64, i64)>,
}

#[inline]
fn parse_hex(s: &str) -> u32 {
    let mut v = 0u32;
    for b in s.as_bytes() {
        v = v * 16
            + match b {
                b'0'..=b'9' => (b - b'0') as u32,
                b'a'..=b'f' => (b - b'a' + 10) as u32,
                _ => (b - b'A' + 10) as u32,
            };
    }
    v
}

#[inline]
fn parse_i64(s: &str) -> i64 {
    let bytes = s.as_bytes();
    let (neg, bytes) = if bytes[0] == b'-' { (true, &bytes[1..]) } else { (false, bytes) };
    let mut v = 0i64;
    for b in bytes {
        v = v * 10 + (b - b'0') as i64;
    }
    if neg {
        -v
    } else {
        v
    }
}

fn parse_rows_owned(text: &str) -> Vec<RowOwned> {
    let mut out = Vec::with_capacity(64 * 1024);
    for line in text.split('\n') {
        if line.is_empty() || line.as_bytes()[0] == b'#' {
            continue;
        }
        let mut it = line.splitn(5, '\t');
        let cps = it.next().unwrap();
        let glyphs = it.next().unwrap();
        let clusters = it.next().unwrap();
        let seams = it.next().unwrap();
        let positions = it.next().unwrap();
        out.push(RowOwned {
            codepoints: cps.split(':').map(parse_hex).collect(),
            glyphs: glyphs.split('|').map(|s| s.to_string()).collect(),
            clusters: clusters.split(',').map(parse_i64).collect(),
            seams: if seams.is_empty() {
                Vec::new()
            } else {
                seams.split(',').map(|s| s.to_string()).collect()
            },
            positions: positions
                .split('|')
                .map(|t| {
                    let mut p = t.splitn(3, ',');
                    (
                        parse_i64(p.next().unwrap()),
                        parse_i64(p.next().unwrap()),
                        parse_i64(p.next().unwrap()),
                    )
                })
                .collect(),
        });
    }
    out
}

fn parse_rows_borrowed(text: &str) -> Vec<RowBorrowed<'_>> {
    let mut out = Vec::with_capacity(64 * 1024);
    for line in text.split('\n') {
        if line.is_empty() || line.as_bytes()[0] == b'#' {
            continue;
        }
        let mut it = line.splitn(5, '\t');
        let cps = it.next().unwrap();
        let glyphs = it.next().unwrap();
        let clusters = it.next().unwrap();
        let seams = it.next().unwrap();
        let positions = it.next().unwrap();
        out.push(RowBorrowed {
            codepoints: cps.split(':').map(parse_hex).collect(),
            glyphs: glyphs.split('|').collect(),
            clusters: clusters.split(',').map(parse_i64).collect(),
            seams: if seams.is_empty() { Vec::new() } else { seams.split(',').collect() },
            positions: positions
                .split('|')
                .map(|t| {
                    let mut p = t.splitn(3, ',');
                    (
                        parse_i64(p.next().unwrap()),
                        parse_i64(p.next().unwrap()),
                        parse_i64(p.next().unwrap()),
                    )
                })
                .collect(),
        });
    }
    out
}

fn canon_row(
    out: &mut Vec<u8>,
    codepoints: &[u32],
    glyphs: &[&str],
    clusters: &[i64],
    seams: &[&str],
    positions: &[(i64, i64, i64)],
) {
    for (i, cp) in codepoints.iter().enumerate() {
        if i > 0 {
            out.push(b':');
        }
        push_u32_hex4(out, *cp);
    }
    out.push(b'\t');
    for (i, g) in glyphs.iter().enumerate() {
        if i > 0 {
            out.push(b'|');
        }
        out.extend_from_slice(g.as_bytes());
    }
    out.push(b'\t');
    for (i, c) in clusters.iter().enumerate() {
        if i > 0 {
            out.push(b',');
        }
        push_i64(out, *c);
    }
    out.push(b'\t');
    for (i, s) in seams.iter().enumerate() {
        if i > 0 {
            out.push(b',');
        }
        out.extend_from_slice(s.as_bytes());
    }
    out.push(b'\t');
    for (i, (x, y, a)) in positions.iter().enumerate() {
        if i > 0 {
            out.push(b'|');
        }
        push_i64(out, *x);
        out.push(b',');
        push_i64(out, *y);
        out.push(b',');
        push_i64(out, *a);
    }
    out.push(b'\n');
}

fn rows_checksum_owned(rows: &[RowOwned]) -> String {
    let mut h = Sha256::new();
    let mut buf = Vec::with_capacity(4096);
    for r in rows {
        buf.clear();
        let glyphs: Vec<&str> = r.glyphs.iter().map(|s| s.as_str()).collect();
        let seams: Vec<&str> = r.seams.iter().map(|s| s.as_str()).collect();
        canon_row(&mut buf, &r.codepoints, &glyphs, &r.clusters, &seams, &r.positions);
        h.update(&buf);
    }
    hex(&h.finalize())
}

fn rows_checksum_borrowed(rows: &[RowBorrowed]) -> String {
    let mut h = Sha256::new();
    let mut buf = Vec::with_capacity(4096);
    for r in rows {
        buf.clear();
        canon_row(&mut buf, &r.codepoints, &r.glyphs, &r.clusters, &r.seams, &r.positions);
        h.update(&buf);
    }
    hex(&h.finalize())
}

// ---- AuditRow ----------------------------------------------------------------------------------

struct AuditOwned {
    config: String,
    codepoints: String,
    kinds: Vec<String>,
    matched_entry: String,
    baseline: Vec<String>,
    new: Vec<String>,
}

fn parse_audit_owned(text: &str) -> Vec<AuditOwned> {
    let mut out = Vec::with_capacity(300_000);
    for (index, line) in text.split('\n').enumerate() {
        if index == 0 || line.is_empty() {
            continue;
        }
        let mut it = line.splitn(6, '\t');
        out.push(AuditOwned {
            config: it.next().unwrap().to_string(),
            codepoints: it.next().unwrap().to_string(),
            kinds: it.next().unwrap().split(',').map(|s| s.to_string()).collect(),
            matched_entry: it.next().unwrap().to_string(),
            baseline: it.next().unwrap().split('|').map(|s| s.to_string()).collect(),
            new: it.next().unwrap().split('|').map(|s| s.to_string()).collect(),
        });
    }
    out
}

fn audit_checksum(rows: &[AuditOwned]) -> String {
    let mut h = Sha256::new();
    let mut buf: Vec<u8> = Vec::with_capacity(1024);
    for r in rows {
        buf.clear();
        buf.extend_from_slice(r.config.as_bytes());
        buf.push(b'\t');
        buf.extend_from_slice(r.codepoints.as_bytes());
        buf.push(b'\t');
        for (i, k) in r.kinds.iter().enumerate() {
            if i > 0 {
                buf.push(b',');
            }
            buf.extend_from_slice(k.as_bytes());
        }
        buf.push(b'\t');
        buf.extend_from_slice(r.matched_entry.as_bytes());
        buf.push(b'\t');
        for (i, b) in r.baseline.iter().enumerate() {
            if i > 0 {
                buf.push(b'|');
            }
            buf.extend_from_slice(b.as_bytes());
        }
        buf.push(b'\t');
        for (i, n) in r.new.iter().enumerate() {
            if i > 0 {
                buf.push(b'|');
            }
            buf.extend_from_slice(n.as_bytes());
        }
        buf.push(b'\n');
        h.update(&buf);
    }
    hex(&h.finalize())
}

// ---- filter_table ------------------------------------------------------------------------------

const ALPHABET: [u32; 18] = [
    0x0020, 0x00B7, 0x200C, 0xE650, 0xE652, 0xE653, 0xE658, 0xE65A, 0xE665, 0xE666, 0xE667, 0xE668,
    0xE670, 0xE672, 0xE675, 0xE676, 0xE679, 0xE67A,
];

#[inline]
fn in_alphabet(token: &[u8]) -> bool {
    let mut v = 0u32;
    for b in token {
        let d = match b {
            b'0'..=b'9' => (b - b'0') as u32,
            b'a'..=b'f' => (b - b'a' + 10) as u32,
            b'A'..=b'F' => (b - b'A' + 10) as u32,
            _ => return false,
        };
        v = v.wrapping_mul(16).wrapping_add(d);
    }
    ALPHABET.contains(&v)
}

#[inline]
fn keep_line(line: &[u8]) -> bool {
    let end = match line.iter().position(|&c| c == b'\t') {
        Some(i) => i,
        None => line.len(),
    };
    let field = &line[..end];
    for token in field.split(|&c| c == b':') {
        if !in_alphabet(token) {
            return false;
        }
    }
    true
}

fn filter_chunk(data: &[u8], out: &mut Vec<u8>) -> usize {
    let mut kept = 0;
    for line in data.split(|&c| c == b'\n') {
        if line.is_empty() {
            continue;
        }
        if line[0] == b'#' {
            out.extend_from_slice(line);
            out.push(b'\n');
            continue;
        }
        if keep_line(line) {
            out.extend_from_slice(line);
            out.push(b'\n');
            kept += 1;
        }
    }
    kept
}

fn split_at_lines(data: &[u8], parts: usize) -> Vec<&[u8]> {
    let mut bounds = Vec::with_capacity(parts + 1);
    bounds.push(0usize);
    for i in 1..parts {
        let mut p = data.len() * i / parts;
        while p < data.len() && data[p] != b'\n' {
            p += 1;
        }
        if p < data.len() {
            p += 1;
        }
        bounds.push(p.min(data.len()));
    }
    bounds.push(data.len());
    bounds.dedup();
    (0..bounds.len() - 1).map(|i| &data[bounds[i]..bounds[i + 1]]).collect()
}

fn hex(bytes: &[u8]) -> String {
    let mut s = String::with_capacity(bytes.len() * 2);
    for b in bytes {
        s.push(char::from_digit((b >> 4) as u32, 16).unwrap());
        s.push(char::from_digit((b & 15) as u32, 16).unwrap());
    }
    s
}

fn best<F: FnMut() -> R, R>(mut f: F, reps: usize) -> (f64, R) {
    let mut b = f64::INFINITY;
    let mut last = None;
    for _ in 0..reps {
        let t = Instant::now();
        let r = f();
        let e = t.elapsed().as_secs_f64();
        if e < b {
            b = e;
        }
        last = Some(r);
    }
    (b, last.unwrap())
}

fn main() {
    let a: Vec<String> = std::env::args().collect();
    let rows_path = &a[1];
    let audit_path = &a[2];
    let big_path = &a[3];
    let out_dir = &a[4];
    let threads: usize = a.get(5).and_then(|s| s.parse().ok()).unwrap_or(8);

    let rows_text = std::fs::read_to_string(rows_path).unwrap();
    let (t_owned, owned) = best(|| parse_rows_owned(&rows_text), 5);
    let n_rows = owned.len();
    let ck_owned = rows_checksum_owned(&owned);
    drop(owned);
    let (t_borrowed, borrowed) = best(|| parse_rows_borrowed(&rows_text), 5);
    let ck_borrowed = rows_checksum_borrowed(&borrowed);
    drop(borrowed);

    let audit_text = std::fs::read_to_string(audit_path).unwrap();
    let (t_audit, audit) = best(|| parse_audit_owned(&audit_text), 3);
    let n_audit = audit.len();
    let ck_audit = audit_checksum(&audit);
    drop(audit);

    let big = std::fs::read(big_path).unwrap();
    let big_len = big.len();
    let (t_filter, (kept1, out1)) = best(
        || {
            let mut out = Vec::with_capacity(8 << 20);
            let kept = filter_chunk(&big, &mut out);
            (kept, out)
        },
        2,
    );
    let ck_filter = hex(&Sha256::digest(&out1));
    std::fs::write(format!("{}/rust.subset.tsv", out_dir), &out1).unwrap();
    drop(out1);

    let (t_filter_par, (kept2, out2)) = best(
        || {
            let chunks = split_at_lines(&big, threads);
            let parts: Vec<(usize, Vec<u8>)> = std::thread::scope(|scope| {
                let handles: Vec<_> = chunks
                    .iter()
                    .map(|chunk| {
                        scope.spawn(move || {
                            let mut out = Vec::with_capacity(2 << 20);
                            let kept = filter_chunk(chunk, &mut out);
                            (kept, out)
                        })
                    })
                    .collect();
                handles.into_iter().map(|h| h.join().unwrap()).collect()
            });
            let mut kept = 0;
            let mut out = Vec::with_capacity(8 << 20);
            for (k, part) in parts {
                kept += k;
                out.extend_from_slice(&part);
            }
            (kept, out)
        },
        2,
    );
    let ck_filter_par = hex(&Sha256::digest(&out2));
    drop(out2);

    println!(
        "{{\"rows_from_tsv\":{{\"rows\":{},\"rust_owned_seconds\":{:.6},\"rust_owned_ns_per_row\":{:.1},\
        \"rust_owned_checksum\":\"{}\",\"rust_borrowed_seconds\":{:.6},\"rust_borrowed_ns_per_row\":{:.1},\
        \"rust_borrowed_checksum\":\"{}\"}},\
        \"load_audit\":{{\"rows\":{},\"rust_seconds\":{:.6},\"rust_ns_per_row\":{:.1},\"rust_checksum\":\"{}\"}},\
        \"filter_table\":{{\"source_bytes\":{},\"rust_single_seconds\":{:.6},\"rust_single_kept\":{},\
        \"rust_single_checksum\":\"{}\",\"rust_parallel_threads\":{},\"rust_parallel_seconds\":{:.6},\
        \"rust_parallel_kept\":{},\"rust_parallel_checksum\":\"{}\"}}}}",
        n_rows,
        t_owned,
        t_owned / n_rows as f64 * 1e9,
        ck_owned,
        t_borrowed,
        t_borrowed / n_rows as f64 * 1e9,
        ck_borrowed,
        n_audit,
        t_audit,
        t_audit / n_audit as f64 * 1e9,
        ck_audit,
        big_len,
        t_filter,
        kept1,
        ck_filter,
        threads,
        t_filter_par,
        kept2,
        ck_filter_par
    );
}
